api/src/main/java/jakarta/data/metamodel/Pattern.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.metamodel;
+
+import jakarta.data.Operator;
+
+
+/**
+ * Represents a pattern-based restriction for matching operations, encapsulating different
+ * options such as prefix, suffix, and substring matching. This implementation
+ * allows flexibility in creating pattern-based conditions directly as `Restriction` instances.
+ *
+ * <p>Example usage with metadata attributes:</p>
+ * <pre>
+ * // Case-sensitive exact match
+ * Restriction<Book> exactMatch = Pattern.like(_Book.title, "Jakarta Data");
+ *
+ * // Case-insensitive prefix match
+ * Restriction<Book> prefixIgnoreCase = Pattern.prefixedIgnoreCase(_Book.title, "Jak");
+ *
+ * // Case-sensitive suffix match
+ * Restriction<Book> suffixMatch = Pattern.suffixed(_Book.title, "Guide");
+ *
+ * // Case-insensitive substring match
+ * Restriction<Book> substringIgnoreCase = Pattern.substringedIgnoreCase(_Book.title, "Java");
+ * </pre>
+ *
+ * @param <T> the type of the entity on which the restriction is applied.
+ */
+public record Pattern<T>(String field, String value, boolean ignoreCase) implements BasicRestriction<T> {
```

Comment:
Don't inherit from BasicRestriction, and don't include `field` inside the Pattern.

Reply:
It seems nonsensical to me to have a restriction that doesn't apply to any entity field in particular.  Having a Pattern be independent of a field makes sense to me. A field isn't part of a pattern.

Reply:
> Using the factory method, the user won't see it at all. Thus, it will also work for our metamodel structure, thus:
> 
> ```java
> Restriction<Book> filter = Pattern.like("...");
> Restriction<Book> filter = _Book.title.like("...");
> Restriction<Book> filter = _Book.title.like(Pattern.like("..."));
> ```
> 
> All of those three will work.

That second and third are good. The first one is an invalid restriction that will fail when used.  A Pattern is text, typically with `%` or `_` wildcards.  A Restriction is the combination of an entity attribute, a comparison operator, information about whether to negate or ignore case, and a value to compare with.  A Pattern is not a Restriction.

Reply:
>  A Pattern is not a Restriction.

That is a good point, on NoSQL to archive it we would use the LIKE operator against the field that the user wish...

So, on JPA we do that without operator? Do you do it on memory instead of returning on the database?

That sounds no efficient.

Reply:
> So, on JPA we do that without operator? Do you do it on memory instead of returning on the database?

No. I don't know why you would conclude that from what I said.

> use the LIKE operator against the field that the user wish...

This part is important. Pattern has no field.  You need the field to come from either a Restriction that is built using the Pattern as its value,

```
Restriction<Book> filter = _Book.title.like(Pattern.startsWith("Jakarta "))
                                      .ignoreCase();
```

or from a repository method indicating which field it pertains to,

```
@Find
List<Book> named(@Is(ANY_CASE) Pattern name);
```

which has all of the same information you could use to construct a Restriction.

Reply:
@otaviojava I agree with Nathan here.

`Pattern` should be a standalone class, not a kind of `Restriction`. It's not a elegant, nor typesafe, to have one field being nullable just completely change the role of a type and what you're supposed to use it for.

Reply:
@njr-11 @gavinking I got your point.
I've moved Pattern to the `Jakarta.data` package and removed the Restriction interface.


Please, let me know what you think and please, feel free to open PR as well.